File: server/controllers/cartController.js
import Cart from "../models/Cart.js";

export const addItemToCart = async (req, res) => {
  const { userId } = req.params;
  const { productId, name, price, quantity, size, imageUrl } = req.body;

  try {
    let cart = await Cart.findOne({ userId });

    if (!cart) {
      cart = new Cart({ userId, items: [] });
    }

    const existingItemIndex = cart.items.findIndex(
      (item) => item.productId === productId && item.size === size
    );

    if (existingItemIndex >= 0) {
      cart.items[existingItemIndex].quantity += quantity;
    } else {
      cart.items.push({ productId, name, price, quantity, size, imageUrl });
    }

    await cart.save();
    res.status(200).json(cart.items);
  } catch (error) {
    res.status(500).json({ message: "Something went wrong" });
  }
};
export const getCart = async (req, res) => {
  const { userId } = req.params;

  try {
    const cart = await Cart.findOne({ userId });
    if (!cart) {
      return res.status(404).json({ message: "Cart not found" });
    }
    res.status(200).json(cart.items);
  } catch (error) {
    res.status(500).json({ message: "Something went wrong" });
  }
};

export const removeItemFromCart = async (req, res) => {
  const { userId, productId } = req.params;
  const { size } = req.body;

  try {
    const cart = await Cart.findOne({ userId });

    if (!cart) {
      return res.status(404).json({ message: "Cart not found" });
    }

    cart.items = cart.items.filter(
      (item) => item.productId !== productId || item.size !== size
    );

    await cart.save();
    res.status(200).json(cart.items);
  } catch (error) {
    res.status(500).json({ message: "Something went wrong" });
  }
};

File: backend-dotnet/ECommerce.Api/Features/Cart/CartController.cs
using ECommerce.Api.Common.Models;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ECommerce.Api.Features.Cart;

[Authorize]
[ApiController]
[Route("api/cart")]
public class CartController : ControllerBase
{
    private readonly IMediator _mediator;
    public CartController(IMediator mediator) => _mediator = mediator;

    /// <summary>Get user's cart</summary>
    [HttpGet("{userId:int}")]
    public async Task<IActionResult> GetCart(int userId)
    {
        var result = await _mediator.Send(new GetCartQuery(userId));
        return result.IsSuccess ? Ok(result.Value) : StatusCode(result.StatusCode, new { message = result.Error });
    }

    /// <summary>Add item to cart</summary>
    [HttpPost("{userId:int}")]
    public async Task<IActionResult> AddToCart(int userId, [FromBody] AddToCartBody body)
    {
        var command = new AddToCartCommand(userId, body.ProductId, body.Name, body.Price, body.Quantity, body.Size, body.ImageUrl);
        var result = await _mediator.Send(command);
        return result.IsSuccess ? Ok(result.Value) : StatusCode(result.StatusCode, new { message = result.Error });
    }

    /// <summary>Remove item from cart</summary>
    [HttpDelete("{userId:int}/{productId:int}")]
    public async Task<IActionResult> RemoveFromCart(int userId, int productId, [FromBody] RemoveFromCartBody? body)
    {
        var command = new RemoveFromCartCommand(userId, productId, body?.Size);
        var result = await _mediator.Send(command);
        return result.IsSuccess ? Ok(result.Value) : StatusCode(result.StatusCode, new { message = result.Error });
    }
}

public record AddToCartBody(int ProductId, string Name, decimal Price, int Quantity, string? Size, string ImageUrl);
public record RemoveFromCartBody(string? Size);
